File: Arrays_Strings_Linked_Lists/Greedy_Problems_Arrays/Maximize_Sum_K_Negations.cpp
#include<bits/stdc++.h>
using namespace std; 

class Solution {
public:
    int largestSumAfterKNegations(vector<int>& nums, int k) {
        priority_queue<int, vector<int>, greater<int>> pq(nums.begin(), nums.end()); // Build heap
        for(int i=0; i<k; i++) {
            int a = pq.top();
            pq.pop();
            pq.push(-a);
        }
        int ans = 0;
        while(!pq.empty()) {
            ans += pq.top();
            pq.pop();
        }
        return ans;
    }
};

File: Arrays_Strings_Linked_Lists/Greedy_Problems_Arrays/Maximize_Sum_K_Negations.py
import heapq
from typing import List

class Solution:
    # T: O(n+klog(n))
    # S: O(1)
    def largestSumAfterKNegations(self, nums: List[int], k: int) -> int:
        heapq.heapify(nums)  # Creates a min heap by default O(n)
        for _ in range(k):   # O(klog(n))
            heapq.heappush(nums, -heapq.heappop(nums))
        return sum(nums) # O(n)